torchao/csrc/cpu/int8_sdpa.cpp

```diff
@@ -0,0 +1,1918 @@
+  auto k = (key.to(at::kFloat) - k_zp) * k_scale;
+  auto v = (value.to(at::kFloat) - v_zp) * v_scale;
+  const auto scaling_factor = calculate_scale(q, scale);
+  auto attn = at::matmul(q, k.transpose(-2, -1)) * scaling_factor;
+  if (attn_mask.has_value() && attn_mask.value().numel()) {
+    attn = attn.add(attn_mask.value().to(at::kFloat));
+  }
+  attn = at::softmax(attn, -1);
+  // quant attn
+  attn = at::clamp_max(
+      at::clamp_min(at::round(attn / a_scale) + a_zp, 0), 255
+  );
+  // dequant attn
+  attn = (attn - a_zp) * a_scale;
+  auto output = at::matmul(attn, v);
+  // quant output
+  output = at::clamp_max(
+      at::clamp_min(at::round(output / o_scale) + o_zp, 0), 255
+  ).to(at::kByte);
+  return output;
+}
+
+
+at::Tensor _scaled_dot_product_int8_cpu(
+    const at::Tensor& query,
+    const at::Tensor& key,
+    const at::Tensor& value,
+    std::optional<at::Tensor> attn_mask,
+    double dropout_p,
+    bool is_causal,
+    double scale,
+    int64_t q_zp,
+    double q_scale,
+    int64_t k_zp,
+    double k_scale,
+    int64_t v_zp,
+    double v_scale,
+    int64_t a_zp,
+    double a_scale,
+    int64_t o_zp,
+    double o_scale) {
+  const auto dtype = query.scalar_type();
+  TORCH_CHECK(!query.is_nested() && !key.is_nested() && !value.is_nested(),
+    "_scaled_dot_product_int8_cpu: Only accept plain inputs");
+  TORCH_CHECK(!is_causal,
+    "_scaled_dot_product_int8_cpu: is_causal not supported.");
+  TORCH_CHECK(dtype == at::ScalarType::Byte,
+    "_scaled_dot_product_int8_cpu: Expected data type be U8, but got ", dtype, " instead.");
+  TORCH_CHECK(query.dim() == 4 && key.dim() == 4 && value.dim() == 4,
+    "_scaled_dot_product_int8_cpu: Accept only 4 dims inputs shape of {B, H, T, K}");
+  TORCH_CHECK(dropout_p == 0.0,
+    "_scaled_dot_product_int8_cpu: Currently do not support dropout > 0");
+  TORCH_CHECK((query.size(3) == value.size(3)) && (key.size(3) == value.size(3)),
+    "_scaled_dot_product_int8_cpu: Q/K/V should have the same head size");
+  TORCH_CHECK(!attn_mask.has_value() ||
+          attn_mask.value().scalar_type() == at::kFloat ||
+          attn_mask.value().scalar_type() == at::kBFloat16,
+    "_scaled_dot_product_int8_cpu: Expected attention mask be float or bf16");
+  TORCH_CHECK(!attn_mask.has_value() ||
+          (attn_mask.value().dim() == 2 || attn_mask.value().dim() == 4),
+    "_scaled_dot_product_int8_cpu: Attention mask dim in {2, 4}");
+
+  #ifdef CPU_CAPABILITY_AVX512
+    if (at::native::cpublas::could_pack(dtype)) {
+        at::Tensor output = at::empty_like(query, query.options()).transpose(1, 2);
+        sdpa_int8_fused_kernel(output, query, key, value,
+            dropout_p, is_causal, attn_mask, scale,
+            q_zp, q_scale,
+            k_zp, k_scale,
+            v_zp, v_scale,
+            a_zp, a_scale,
+            o_zp, o_scale);
+        return output.transpose(1, 2);
+    } else {
+        return sdpa_int8_math_kernel(query, key, value,
```

Comment:
Feels we can merge these 2 `sdpa_int8_math_kernel`.

Reply:
Thanks, modified!